src/herder/QuorumIntersectionCheckerImpl.cpp

```diff
@@ -461,10 +462,31 @@ QuorumIntersectionCheckerImpl::containsQuorumSliceForNode(BitSet const& bs,
     return containsQuorumSlice(bs, mGraph.at(node));
 }
 
+size_t
+QuorumIntersectionCheckerImpl::BitSetHashFunction::
+operator()(const BitSet& nodes) const
+{
+    std::vector<bool> v(nodes.size());
+    for (size_t i = 0; nodes.nextSet(i); ++i)
+    {
+        v[i] = 1;
+    }
+    return hashFunction(v);
+}
+
 bool
 QuorumIntersectionCheckerImpl::isAQuorum(BitSet const& nodes) const
 {
-    return (bool)contractToMaximalQuorum(nodes);
+    if (!cachedQuorums.exists(nodes))
```

Comment:
This is performing multiple lookups in the table all the time and is likely high impact given the number of time we hit this.

There is mention of `maybeGet` in `RandomEvictionCache` but it looks like it was removed/never implemented. So we should just add it (it's just a small refactor of `get`) and use it instead.

Code will look like this:
```
    auto v = cachedQuorums.maybeGet(nodes);
    if (!v)
    {
      // this will also recompute the hash
    }
    else
    {
         return *v;
    }
```

This will at least fix the "cached" path, the "put" path is a bit trickier to fix (and not worth it for now)

Reply:
I implemented this, but I was not able to get the performance improvements that I hoped from this.

https://github.com/hidenori-shinohara/stellar-core/commit/0a8c00a57a2dde8ba96c65cdbbe8042b684bbf3c

Without `maybeGet`, the code was using
```
bytes allocated in total (ignoring deallocations): 24.67MB (7.43MB/s)
calls to allocation functions: 277527 (83592/s)
temporary memory allocations: 1736 (522/s)
peak heap memory consumption: 8.17MB
peak RSS (including heaptrack overhead): 292.10MB
total memory leaked: 1.28MB
```
, but with `maybeGet`, the code used

```
bytes allocated in total (ignoring deallocations): 31.31MB (10.03MB/s)                                                                                                                                                                         
calls to allocation functions: 588701 (188625/s)                                                                                                                                                                                               
temporary memory allocations: 327594 (104964/s)                                                                                                                                                                                                
peak heap memory consumption: 8.17MB                                                                                                                                                                                                           
peak RSS (including heaptrack overhead): 310.98MB                                                                                                                                                                                              
total memory leaked: 1.28MB
```

I'm wondering if there is a way to improve the implementation. The only other hypothesis that I could come up with was that this is because every return value (which was just a boolean that takes 1 byte) became `optional<bool>` which seems to take 16 bytes when I used `sizeof`.

Reply:
re: `maybeGet` you can return a pointer instead of an optional and avoid extra allocations

Reply:
with the change done like this, you should also change `get` to use `maybeGet` to avoid code duplication

Reply:
This indeed improved both the performance and bytes allocated in total by ~10% ish!